### PR TITLE
Create amazonlinux2023 template

### PR DESCRIPTION
**Pull request by siennathesane**
Adds a default template reference for Amazon Linux 2023.

### CODE REVIEW

**Comment by jandubois** [comment]
Please fix lint error (remove trailing spaces on first line) and apply DCO signature to the commit: https://github.com/lima-vm/lima/pull/2537/checks?check_run_id=28484568245

File: examples/amazonlinux2023.yaml
Comment (AkihiroSuda) [inline examples/amazonlinux2023.yaml:2]: Why is the image incompatible with qemu?

File: examples/amazonlinux2023.yaml
Comment (AkihiroSuda) [inline examples/amazonlinux2023.yaml:1]: The filename should be amazonlinux.yaml

File: examples/amazonlinux2023.yaml
Comment (AkihiroSuda) [inline examples/amazonlinux2023.yaml:5]: Are these images persistent for years, or are they wiped periodically?
In the latter case, the "latest" image has to be specified as a fallback option.
(See other templates)

**Comment by afbjorklund** [comment]
There is a https://cdn.amazonlinux.com/al2023/os-images/latest/ link, but all filenames seem to be unique?

https://docs.aws.amazon.com/linux/al2023/ug/outside-ec2-download.html

Maybe there could be an issue or documentation created upstream, how to link to the latest KVM image?

https://aws.amazon.com/about-aws/whats-new/2023/12/amazon-linux-kvm-vmware-images-al2023-3/

---

* https://aws.amazon.com/linux/amazon-linux-2023/

* https://github.com/amazonlinux/amazon-linux-2023

Maybe [finch](https://aws.amazon.com/blogs/opensource/introducing-finch-an-open-source-client-for-container-development/) wants to help out, as well ?

Or I suppose you _could_ just use `fedora` :-)

File: examples/amazonlinux2023.yaml
Comment (siennathesane) [reply examples/amazonlinux2023.yaml:1]: I disagree because Amazon Linux, Amazon Linux 2, and Amazon Linux 2023 are different operating systems in the same way that Fedora & RedHat are different operating systems. Given #625 was denied, it's an important distinction.

File: examples/amazonlinux2023.yaml
Comment (siennathesane) [reply examples/amazonlinux2023.yaml:5]: > Are these images persistent for years, or are they wiped periodically?

I don't know, that's an Amazon question, sorry. To my knowledge there is no `latest` option.

**Comment by siennathesane** [comment]
> There is a https://cdn.amazonlinux.com/al2023/os-images/latest/ link, but all filenames seem to be unique?
> 
> https://docs.aws.amazon.com/linux/al2023/ug/outside-ec2-download.html
> 
> Maybe there could be an issue or documentation created upstream, how to link to the latest KVM image?
> 
> https://aws.amazon.com/about-aws/whats-new/2023/12/amazon-linux-kvm-vmware-images-al2023-3/

Do you want a link to that in the example file? I think anything more than that is more than I want to do for the scope of this PR.
 
> Maybe [finch](https://aws.amazon.com/blogs/opensource/introducing-finch-an-open-source-client-for-container-development/) wants to help out, as well ?

The two are different projects, I'm not sure how they relate, sorry.

> Or I suppose you _could_ just use `fedora` :-)

Amazon Linux 2023 & Fedora are not the same despite many similarities.

File: examples/amazonlinux2023.yaml
Comment (siennathesane) [reply examples/amazonlinux2023.yaml]: I misunderstood the documentation. Fixed.

**Comment by siennathesane** [comment]
I moved it into experimental to allow for more user awareness of the state due to lack of `latest` support.

File: examples/amazonlinux2023.yaml
Comment (AkihiroSuda) [reply examples/amazonlinux2023.yaml:1]: I still don't get why the new distro (and later versions) can't be called "amazonlinux".

Does that mean that the new distro will be called "Amazon Linux 2023" practically forever, even in 2030s?

Comment (jandubois) [reply examples/amazonlinux2023.yaml:1]: No, there will be "Amazon Linux 2025", "Amazon Linux 2027", etc in the future.

You can see the expected lifecycles at https://docs.aws.amazon.com/linux/al2023/ug/release-cadence.html

Comment (jandubois) [reply examples/amazonlinux2023.yaml:1]: > No, there will be "Amazon Linux 2025", "Amazon Linux 2027", etc in the future.

So given that in 2027 there will be AL2023 and AL2025 in maintenance, and AL2027 will be the current version, I think it makes sense to include the "2023" in the name.

I wonder though if it should be `amazonlinux-2023.yaml`, and then have an `amazonlinux.yaml` symlink to it, similarly to how we have `almalinux-9.yaml` or `debian-12.yaml`. That way in the future we can symlink to `amazonlinux-2025.yaml` while still keeping the old `amazonlinux-2023.yaml` around until it is EOL.

File: examples/amazonlinux2023.yaml
Comment (jandubois) [reply examples/amazonlinux2023.yaml]: > Are these images persistent for years, or are they wiped periodically?

I think "somebody" will have to reach out to AWS to find out. I'll be offline soon, so I can't do it right now; ping me in September if it is still unresolved by then. 😄 

Maybe ask the `finch` people to find out who can help. They have a channel on the CNCF Slack.

**Comment by afbjorklund** [comment]
I was just referring to that Amazon projects (like linux or finch) could make this easier for lima?

The template itself doesn't need that information (at the moment), beyond the "latest" image.

File: examples/amazonlinux2023.yaml
Comment (afbjorklund) [reply examples/amazonlinux2023.yaml:1]: And I think that amazon linux 2 (amazonlinux-2.yaml ?) was mostly denied because it was too old, it _could_ have gone into deprecated next to centos-7.yaml (now dead). Or just wait for the new ~~2022~~ 2023 edition (this).

File: examples/experimental/amazonlinux2023.yaml
Comment (afbjorklund) [inline examples/experimental/amazonlinux2023.yaml:2]: this seems a bit out of place, and won't always work on qemu

if you want to change the defaults, it could in _config instead?

File: examples/amazonlinux2023.yaml
Comment (afbjorklund) [reply examples/amazonlinux2023.yaml]: It _could_ be left to just go 404 when it is removed, like with the other distributions?

The "latest" is a nice to have / recovery, but ultimately they all need bumping anyway.

**Comment by pendo324** [comment]
> Maybe there could be an issue or documentation created upstream, how to link to the latest KVM image?

Unfortunately, I think this is the best we've got right now:

```
# extract the latest release from the repo
# append .0 to it so that it works properly in the CDN URL
# I'm not sure if this always has to be .0 or if it can be other digits
distribution_release_al2023="$(curl -s https://al2023-repos-us-west-2-de612dc2.s3.dualstack.us-west-2.amazonaws.com/core/releasemd.xml | xmllint --xpath "string(//root/releases/release[last()]/@version)" -).0"

# extract the filename from the SHASUMS file since that file name is "static"
filename=$(curl -Ls https://cdn.amazonlinux.com/al2023/os-images/2023.5.20240805.0/kvm/SHA256SUMS | awk -F '[[:space:]]+' '{print $2}')

curl -OL https://cdn.amazonlinux.com/al2023/os-images/$distribution_release_al2023/kvm/$filename
```

File: examples/amazonlinux2023.yaml
Comment (jandubois) [reply examples/amazonlinux2023.yaml:1]: Looks like we have agreement then.

I've received confirmation that the images are supposed to live "forever", so I think we don't need a fallback URL for the "latest" version.

@siennathesane please rename `amazonlinux2023.yaml` to `amazonlinux-2023.yaml`.

I guess we can wait with creating an `amazonlinux.yaml` symlink until the template has graduated from "experimental" to a regular template.

Comment (afbjorklund) [reply examples/amazonlinux2023.yaml:1]: I don't think it has to go in experimental either, just because upstream is missing a symlink?

**Comment by afbjorklund** [comment]
The update procedure is a separate topic (and not a requirement), many others also lack a good method.

* https://github.com/lima-vm/lima/issues/1347

File: examples/experimental/amazonlinux2023.yaml
Comment (siennathesane) [reply examples/experimental/amazonlinux2023.yaml:2]: This is the only supported config: https://docs.aws.amazon.com/linux/al2023/ug/kvm-supported-configurations.html

File: examples/amazonlinux2023.yaml
Comment (siennathesane) [reply examples/amazonlinux2023.yaml:1]: I put it in experimental due to no `latest` version and because it requires `virtiofs`. Since `virtiofs` requires experimental `vz` on macOS, it made more sense. Do you want me to move it?

Comment (siennathesane) [reply examples/amazonlinux2023.yaml:1]: I put it in experimental due to no `latest` version and because it requires `virtiofs`. Since `virtiofs` requires experimental `vz` on macOS, it made more sense. Do you want me to move it?

File: examples/README.md
Comment (jandubois) [inline examples/README.md:17]: The file has been renamed:

```suggestion
- [`amazonlinux-2023.yaml`](./experimental/amazonlinux-2023.yaml): [experimental] Amazon Linux 2023
```

**Review by jandubois — CHANGES_REQUESTED** [review]
Thanks, LGTM, but the file rename needs to be reflected in the README.

**Comment by jandubois** [comment]
You are missing the DCO sign-off. Also please squash commits!

File: examples/experimental/amazonlinux-2023.yaml
Comment (afbjorklund) [inline examples/experimental/amazonlinux-2023.yaml:2]: I don't think the mountType should be in here.

File: examples/amazonlinux2023.yaml
Comment (afbjorklund) [reply examples/amazonlinux2023.yaml:1]: Does it require virtiofs, or does it just not report 9p?

File: examples/experimental/amazonlinux2023.yaml
Comment (afbjorklund) [reply examples/experimental/amazonlinux2023.yaml]: Just wondered why it was needed here, but not in any other template

Comment (afbjorklund) [reply examples/experimental/amazonlinux2023.yaml]: It was needed because Amazon Linux does not provide any `sshfs` rpm

File: examples/amazonlinux2023.yaml
Comment (afbjorklund) [reply examples/amazonlinux2023.yaml:1]: Also, `# CONFIG_NET_9P is not set` in kernel. So that means that it does indeed only support `virtiofs`.

File: examples/experimental/amazonlinux-2023.yaml
Comment (afbjorklund) [reply examples/experimental/amazonlinux-2023.yaml:2]: Both other mount types are broken on Amazon Linux, so the only alternative is removing default mounts...

**Comment by siennathesane** [comment]
> You are missing the DCO sign-off. Also please squash commits!

Done

File: examples/experimental/amazonlinux-2023.yaml
Comment (jandubois) [reply examples/experimental/amazonlinux-2023.yaml:2]: Given that the PR needs another change anyways, I think it would be good to add a comment here that `reverse-sshfs`, and `9p` are not supported by amazonlinux, so people know not to bother trying.

**Review by jandubois — CHANGES_REQUESTED** [review]
Thank you! Unfortunately not there yet:

* The filenames in the README need to be updated to match the filenames on disk.

* Adding a comment about the `mountType` would be nice.

* Commits need to be squashed.

**Comment by siennathesane** [comment]
> Commits need to be squashed.

Sorry, I didn't realise y'all use `master` so `git reset --soft HEAD~$(git rev-list --count HEAD ^main)` didn't catch everything. Addressed & fixed.

File: examples/experimental/amazonlinux-2023.yaml
Comment (siennathesane) [reply examples/experimental/amazonlinux-2023.yaml:3]: fixed

**Review by jandubois — APPROVED** [review]
Thanks!

PR looks good to me now, but while testing it on Sonoma with M1 I noticed that the mounts didn't actually work. Do they work for anybody else?

**Comment by AkihiroSuda** [comment]
@siennathesane Could you take a look at the mount issue?

**Comment by siennathesane** [comment]
> @siennathesane Could you take a look at the mount issue?

This PR has been enough effort just to share an experimental example. I think it's unclear to me what the purpose of examples are if there's such a high bar to share something that's also experimental. The intent of this PR was to share an example of what an AWS config could look like, because it's my understanding that these are _examples_, esp _experimental_ ones.

At this point if there's a reason not to merge, then go ahead and close the PR, but I'm not going to spend any more time on this.

**Comment by AkihiroSuda** [comment]
> This PR has been enough effort just to share an experimental example. I think it's unclear to me what the purpose of examples are if there's such a high bar to share something that's also experimental. The intent of this PR was to share an example of what an AWS config could look like, because it's my understanding that these are examples, esp experimental ones.

"examples" have been renamed to "templates" in https://github.com/lima-vm/lima/pull/1679 to clarify that they are expected to be useful OOTB, but the git repo still uses `examples` so as to keep hyperlinks alive.

The templates, especially experimental ones, may still have non-functional features.
However, such a non-functionality has to be explained in the comment lines of the yaml.

**Comment by kikokikok** [comment]
I've hit the same issue doing it from scratch before landing here...
could someone help me with where to look for ?
so far i've been able to see that 
``` bash
Cloud-init v. 22.2.2 running 'modules:config' at Mon, 14 Oct 2024 16:15:35 +0000. Up 9.95 seconds.
Cloud-init v. 22.2.2 running 'modules:final' at Mon, 14 Oct 2024 16:15:35 +0000. Up 10.15 seconds.
+ LIMA_CIDATA_MNT=/mnt/lima-cidata
+ LIMA_CIDATA_DEV=/dev/disk/by-label/cidata
+ mkdir -p -m 700 /mnt/lima-cidata
+ mount -o ro,mode=0700,dmode=0700,overriderockperm,exec,uid=0 /dev/disk/by-label/cidata /mnt/lima-cidata
+ export LIMA_CIDATA_MNT
+ exec /mnt/lima-cidata/boot.sh
LIMA 2024-10-14T18:15:35+02:00| Executing /mnt/lima-cidata/boot/00-alpine-user-group.sh
LIMA 2024-10-14T18:15:35+02:00| Executing /mnt/lima-cidata/boot/00-modprobe.sh
Loading kernel module "fuse"
Loading kernel module "tun"
Loading kernel module "tap"
Loading kernel module "bridge"
Loading kernel module "veth"
Loading kernel module "ip_tables"
Loading kernel module "ip6_tables"
Loading kernel module "iptable_nat"
Loading kernel module "ip6table_nat"
Loading kernel module "iptable_filter"
Loading kernel module "ip6table_filter"
Loading kernel module "nf_tables"
Loading kernel module "x_tables"
Loading kernel module "xt_MASQUERADE"
Loading kernel module "xt_addrtype"
Loading kernel module "xt_comment"
Loading kernel module "xt_conntrack"
Loading kernel module "xt_mark"
Loading kernel module "xt_multiport"
Loading kernel module "xt_nat"
Loading kernel module "xt_tcpudp"
Loading kernel module "overlay"
LIMA 2024-10-14T18:15:35+02:00| Executing /mnt/lima-cidata/boot/00-reboot-if-required.sh
+ command -v dnf
+ dnf -h needs-restarting
+ dnf needs-restarting -r
LIMA 2024-10-14T18:15:36+02:00| Executing /mnt/lima-cidata/boot/01-alpine-ash-as-bash.sh
LIMA 2024-10-14T18:15:36+02:00| Executing /mnt/lima-cidata/boot/02-wsl2-setup.sh
LIMA 2024-10-14T18:15:36+02:00| Executing /mnt/lima-cidata/boot/04-persistent-data-volume.sh
+ test -f /etc/alpine-release
+ exit 0
LIMA 2024-10-14T18:15:36+02:00| Executing /mnt/lima-cidata/boot/05-lima-disks.sh
+ test 0 -gt 0
+ exit 0
LIMA 2024-10-14T18:15:36+02:00| Executing /mnt/lima-cidata/boot/05-lima-mounts.sh
+ [[ vz == \v\z ]]
+ [[ virtiofs == \v\i\r\t\i\o\f\s ]]
+ '[' -d /sys/fs/selinux ']'
++ grep -n virtiofs
++ cut -d: -f1
LIMA 2024-10-14T18:15:36+02:00| Executing /mnt/lima-cidata/boot/05-rosetta-volume.sh
+ '[' true '!=' true ']'
+ '[' -f /etc/alpine-release ']'
+ binfmt_entry=/proc/sys/fs/binfmt_misc/rosetta
+ binfmtd_conf=/usr/lib/binfmt.d/rosetta.conf
+ '[' true = true ']'
```

I have tried adding

``` yaml
provision:
- mode: boot
  script: |
    # Load the virtiofs module at boot
    modprobe virtiofs
    # Ensure the module loads on every boot by adding it to the modules-load.d directory
    echo "virtiofs" > /etc/modules-load.d/virtiofs.conf

    #disable SELinux
    grubby --update-kernel ALL --args selinux=0
```

but unfortunately, it seems like the boot step is executed after the mount ? 
Sorry my I don't have the knowledge depth needed to get this done. Willing to help though.

Thanks for any potential advice from someone.

**Comment by kikokikok** [comment]
Small update here 

``` bash
2024-10-14 16:55:05,857 - cc_mounts.py[DEBUG]: mounts configuration is [['vz-rosetta', '/mnt/lima-rosetta', 'virtiofs', 'defaults', '0', '0'], ['mount0', '/Users/christian.klat', 'virtiofs', 'rw,nofail', '0', '0'], ['mount1', '/tmp/lima', 'virtiofs', 'rw,nofail', '0', '0']]
2024-10-14 16:55:05,860 - util.py[DEBUG]: backgrounded Resizing took 0.006 seconds
2024-10-14 16:55:05,857 - util.py[DEBUG]: Reading from /etc/fstab (quiet=False)
2024-10-14 16:55:05,863 - util.py[DEBUG]: Read 217 bytes from /etc/fstab
2024-10-14 16:55:05,863 - cc_mounts.py[DEBUG]: Attempting to determine the real name of vz-rosetta
2024-10-14 16:55:05,863 - cc_mounts.py[DEBUG]: changed vz-rosetta => None
2024-10-14 16:55:05,863 - cc_mounts.py[DEBUG]: Ignoring nonexistent named mount vz-rosetta
2024-10-14 16:55:05,863 - cc_mounts.py[DEBUG]: Attempting to determine the real name of mount0
2024-10-14 16:55:05,863 - cc_mounts.py[DEBUG]: changed mount0 => None
2024-10-14 16:55:05,863 - cc_mounts.py[DEBUG]: Ignoring nonexistent named mount mount0
2024-10-14 16:55:05,863 - cc_mounts.py[DEBUG]: Attempting to determine the real name of mount1
2024-10-14 16:55:05,863 - cc_mounts.py[DEBUG]: changed mount1 => None
2024-10-14 16:55:05,863 - cc_mounts.py[DEBUG]: Ignoring nonexistent named mount mount1
2024-10-14 16:55:05,863 - cc_mounts.py[DEBUG]: Attempting to determine the real name of ephemeral0
2024-10-14 16:55:05,864 - cc_mounts.py[DEBUG]: changed default device ephemeral0 => None
2024-10-14 16:55:05,864 - cc_mounts.py[DEBUG]: Ignoring nonexistent default named mount ephemeral0
2024-10-14 16:55:05,864 - cc_mounts.py[DEBUG]: Attempting to determine the real name of swap
2024-10-14 16:55:05,864 - cc_mounts.py[DEBUG]: changed default device swap => None
2024-10-14 16:55:05,864 - cc_mounts.py[DEBUG]: Ignoring nonexistent default named mount swap
2024-10-14 16:55:05,864 - cc_mounts.py[DEBUG]: Skipping nonexistent device named None
2024-10-14 16:55:05,864 - cc_mounts.py[DEBUG]: Skipping nonexistent device named None
2024-10-14 16:55:05,864 - cc_mounts.py[DEBUG]: Skipping nonexistent device named None
2024-10-14 16:55:05,864 - cc_mounts.py[DEBUG]: no need to setup swap
2024-10-14 16:55:05,864 - cc_mounts.py[DEBUG]: No modifications to fstab needed

```
Seems like the cc_mounts.py is failing at performing the mounts.
- In this run I was able to make sure virtiofs is loaded a reboot happened and then we have this issue.

File: examples/amazonlinux2023.yaml
Comment (electricalgorithm) [reply examples/amazonlinux2023.yaml]: You can see that links are preserved: https://cdn.amazonlinux.com/al2023/os-images/2023.2.20231113.0/ (AL2023 release from 2023-11-13). However, one can use https://cdn.amazonlinux.com/al2023/os-images/latest/ to be sure.

**Comment by electricalgorithm** [comment]
> > Maybe there could be an issue or documentation created upstream, how to link to the latest KVM image?
> 
> Unfortunately, I think this is the best we've got right now:
> 
> ```
> # extract the latest release from the repo
> # append .0 to it so that it works properly in the CDN URL
> # I'm not sure if this always has to be .0 or if it can be other digits
> distribution_release_al2023="$(curl -s https://al2023-repos-us-west-2-de612dc2.s3.dualstack.us-west-2.amazonaws.com/core/releasemd.xml | xmllint --xpath "string(//root/releases/release[last()]/@version)" -).0"
> 
> # extract the filename from the SHASUMS file since that file name is "static"
> filename=$(curl -Ls "https://cdn.amazonlinux.com/al2023/os-images/$distribution_release_al2023/kvm/SHA256SUMS" | awk -F '[[:space:]]+' '{print $2}')
> 
> curl -OL https://cdn.amazonlinux.com/al2023/os-images/$distribution_release_al2023/kvm/$filename
> ```

It's possible to have .1's or .2's, although it's unlikely. It'd be better to stick with a dynamic approach where we can test to get 1,2,3 (never seen a minor version greater than 2).

[Update] A better approach is possible by using the redirection of latest endpoint:
```bash
# Get the latest release version from the CDN redirect
distribution_release_al2023=$(basename "$(curl -Ls -o /dev/null -w %{url_effective} https://cdn.amazonlinux.com/al2023/os-images/latest/)")

echo "Latest Release: ${distribution_release_al2023}"

# Function to get URL and Digest
function get_image_details() {
	local arch="$1"
	local folder="$2" # kvm or kvm-arm64
	local base_url="https://cdn.amazonlinux.com/al2023/os-images/${distribution_release_al2023}/${folder}"
	
	local checksums
	checksums=$(curl -Ls "${base_url}/SHA256SUMS")
	
	local filename
	filename=$(echo "${checksums}" | awk '{print $2}')
	local digest
	digest=$(echo "${checksums}" | awk '{print $1}')
	
	echo "${base_url}/${filename} ${digest}"
}

# Get x86_64 details
read -r url_amd64 digest_amd64 <<< "$(get_image_details "x86_64" "kvm")"
echo "x86_64: ${url_amd64} ${digest_amd64}"

# Get aarch64 details
read -r url_arm64 digest_arm64 <<< "$(get_image_details "aarch64" "kvm-arm64")"
echo "aarch64: ${url_arm64} ${digest_arm64}"
```

**Comment by AkihiroSuda** [comment]
Linking the new PR
- https://github.com/lima-vm/lima/pull/4454

**Pull request closed without merging**
